cryptoprotocol/src/test/java/org/cryptimeleon/incentivesystem/cryptoprotocol/ProtocolIntegrationTest.java

```diff
@@ -0,0 +1,195 @@
+
+        // serialize and deserialize earn request to ensure serialization does not break anything
+        var serializedEarnRequest1 = earnRequest1.getRepresentation();
+        var deserializedEarnRequest1 = new EarnRequest(serializedEarnRequest1, incSys.getPp());
+
+        // provider handles earn request and generates earn response
+        var earnResponse1 = incSys.generateEarnRequestResponse(deserializedEarnRequest1, new BigInteger("20"), pkp);
+
+        // serialize and deserialize earn response to ensure serialization does not break anything
+        var serializedEarnResponse1 = earnResponse1.getRepresentation();
+        var deserializedEarnResponse1 = new SPSEQSignature(serializedEarnResponse1, incSys.getPp().getBg().getG1(), incSys.getPp().getBg().getG2());
+
+        // user handles earn response
+        var updatedToken = incSys.handleEarnRequestResponse(deserializedEarnRequest1, deserializedEarnResponse1, new BigInteger("20"), initialToken, pkp.getPk(), ukp);
+
+        // ensure user token contains 20 points
+        Assertions.assertEquals(updatedToken.getPoints().getInteger(), new BigInteger(("20")));
+
+        /**
+         * transaction 3: user tries to spend 23 points
+         */
+
+        System.out.println("Testing failing spend transaction with non-empty token.");
+
+        // generate a fresh ID for the spend transaction
+        var tid2 = usedZn.getUniformlyRandomElement();
+
+        // define spend amount
+        var spendAmount2 = new BigInteger("23");
+
+        // user tries to generate spend request
+        try
+        {
+            var spendRequest2 = incSys.generateSpendRequest(updatedToken, pkp.getPk(), spendAmount2, ukp, tid2);
+        }
+        catch(IllegalArgumentException e)
+        {
+            System.out.println("Expected exception when trying to spend " + spendAmount2.toString() +" point using token with " + updatedToken.getPoints().toString() + " points: " + e.getMessage());
+        }
+
+        // ensure user token still contains 20 points
+        Assertions.assertEquals(updatedToken.getPoints().getInteger(), new BigInteger(("20")));
+
+        /**
+         * transaction 4: user spends 18 points
+         */
+
+        System.out.println("Testing successful spend transaction.");
+
+        // generate a fresh ID for the spend transaction
+        var tid3 = usedZn.getUniformlyRandomElement();
+
+        // define spend amount
+        var spendAmount3 = new BigInteger("18");
+
+        // user generates spend request
+        SpendRequest spendRequest3 = null;
+        try
+        {
+            spendRequest3 = incSys.generateSpendRequest(updatedToken, pkp.getPk(), spendAmount3, ukp, tid3);
+        }
+        catch(IllegalArgumentException e)
+        {
+            System.out.println("Unexpected exception when trying to spend " + spendAmount3.toString() +" point using token with " + updatedToken.getPoints().toString() + " points: " + e.getMessage());
+        }
```

Comment:
What behavior are you expecting here? You either want the exception or no exception. In the latter case remove the try-catch, in the first assert that there is an exception and **otherwise fail**.

Reply:
Done